Control.Resources, StaticResource and DynamicResource

grokys opened the pull request:
This PR adds `Control.Resources`, `StaticResource` and `DynamicResource`.

The new features are implemented to follow WPF/UWP as closely as possible. The main difference is that you can't do `<ResourceDictionary Source="blah.xaml"/>` in XAML is as we need to be markup-independent. Instead you need to do `<ResourceInclude Source="blah.xaml"/>`. Resource dictionaries are also now keyed on `object` for parity with WPF/UWP.

`StyleResource` should now be removed, but not in this PR - it's still used in our default theme and samples etc. (I didn't want to bloat the PR with changing those to `DynamicResource`s, though everything appears to work. This will be addressed in a subsequent PR). @danwalmsley @wieslawsoltes I don't know how disruptive that will be for you, we can do it later rather than sooner if it will be a problem.

This also fixes all the issues I could find related to resources:

Fixes #492 
Fixes #818
Fixes #1020 
Fixes #1088

grokys committed:
StyleResources -> ResourceDictionary.

And move it to the `Avalonia.Controls` namespace.

grokys committed:
Added IResourceDictionary.

grokys committed:
Remove Style.Resources setter.

We don't need it.

grokys committed:
Made a start adding Control/Application.Resources.

grokys committed:
IResourceHost -> IResourceProvider.

grokys committed:
Fix build failure.

grokys committed:
Fix faulty resource lookup.

grokys committed:
Started adding StaticResource.

grokys committed:
Get Appplication.Resources from StaticResource.

grokys committed:
Delay load static resources if necessary.

grokys committed:
Throw an exception if static resource not found.

grokys committed:
More work on static resources.

Also found a scenario that Portable.Xaml can't seem to handle...

grokys committed:
A few tweaks.

grokys committed:
Moved StaticResourceTests.

grokys committed:
Added DynamicResourceExtension.

grokys committed:
Added more tests.

Some failing.

grokys committed:
Use an anchor for dynamic resources.

grokys committed:
Renamed test classes.

grokys committed:
Use AvaloniaDictionary for ResourceDictionary.

grokys committed:
Start making DynamicResource react to changes.

grokys committed:
Detect resources added in new styles.

grokys committed:
Make dynamic resources work in more situations.

Fixes #492 in a fashion: `DynamicResource` now works for this scenario.

grokys committed:
Update dynamic resources on set parent.

grokys committed:
More resource extension tests.

One failing.

grokys committed:
Ensure properties registered.

When calling `AvaloniaPropertyRegistry.GetAttached`.

grokys committed:
Added tests for #1020

Passing. Closes #1020.

grokys committed:
Static resource as binding converter

Enable using a `StaticResource` as a binding converter. Fixes #818.

grokys committed:
Merge branch 'master' into control-resources

grokys committed:
Updated ncrunch config

grokys committed:
Started adding MergedDictionaries.

grokys committed:
IResourceProvider -> IResourceNode

grokys committed:
Doc comments.

grokys committed:
Start testing MergedDictionaries.

grokys committed:
DynamicResource merged dictionary tests.

grokys committed:
Added ResourceInclude.

grokys committed:
React to application resources changing.

grokys committed:
Make resource dictionary keys objects.

For parity with other XAML frameworks.

grokys committed:
Fixed malformed doc comment.

grokys committed:
Fix compile error.

grokys committed:
Merge branch 'master' into control-resources

 Conflicts:
	src/Avalonia.Controls/Application.cs
	src/Avalonia.Controls/Control.cs
	src/Avalonia.Controls/IControl.cs
	src/Avalonia.Styling/Styling/StyleExtensions.cs
	tests/Avalonia.Styling.UnitTests/ResourceTests.cs

grokys committed:
Merge branch 'master' into control-resources

aelij commented on src/Avalonia.Styling/Styling/Style.cs:
```diff
@@ -112,25 +120,11 @@ public void Attach(IStyleable control, IStyleHost container)
-        /// </returns>
-        public object FindResource(string name)
+        /// <inheritdoc/>
+        public bool TryGetResource(string key, out object result)
```
I think it's a bit confusing to use a method named `FindResource` (with or without Try) both for finding items in a specific node vs. finding them up the hierarchy. WPF made this distinction - Find was for the hierarchy, and you could use the dictionary's indexer or `Resources.TryGetValue` to search a specific node. So I vote to eliminate this method, and use the `Resources` property.

grokys replied:
Not sure what you mean here, `TryGetResource` calls `_resource.TryGetResource` which searches the resource dictionary and all merged dictionaries.

aelij replied:
Yes, but it doesn't search the hierarchy (e.g. control > control > app). In WPF, to access the resources of any item in the dictionary, including the merged ones, you use the `control.Resources["MyKey"]`. To search up the hierarchy, you use `control.TryFindResource`.

aelij commented on src/Avalonia.Styling/Controls/ResourceProviderExtensions.cs:
```diff
@@ -0,0 +1,49 @@
+        /// <param name="control">The control.</param>
+        /// <param name="key">The resource key.</param>
+        /// <returns>The resource, or <see cref="AvaloniaProperty.UnsetValue"/> if not found.</returns>
+        public static object FindResource(this IResourceNode control, string key)
```
Is it possible to change this to `bool TryFindResource(this IResourceNode control, string key, out object value)` and not use `UnsetValue`? The latter was a source of some bugs in WPF, IIRC.

grokys replied:
Yeah, I was considering that but decided to go with the WPF API. I do agree that a `TryGetResource` API would be better though.

grokys replied:
I added `TryFindResource` but for some reason it's not showing up on GH, it's definitely there in the branch though!

aelij commented on src/Markup/Avalonia.Markup.Xaml/MarkupExtensions/StaticResourceExtension.cs:
```diff
@@ -0,0 +1,84 @@
+            if (provideTarget.TargetObject is IControl target &&
+                provideTarget.TargetProperty is PropertyInfo property)
+            {
+                DelayedBinding.Add(target, property, GetValue);
```
Why is this static? Could it be added to the `context`?

grokys replied:
Why is `DelayedBinding` static? Just because I didn't see any reason for it not to be static when I wrote it ;) I don't see how it could be added to the `context`. Could you give more detail on what you're thinking?

aelij replied:
Why not? I see `ApplyBindings` is only called from `AvaloniaXamlLoaderPortableXaml` after `Load`.

grokys replied:
Ok, I'm still not sure what you mean entirely, but considering that `DelayedBinding` is an existing concept, we should probably discuss it in an issue, if you want to open one with more details.

grokys commented on src/Avalonia.Base/Collections/AvaloniaDictionary.cs:
```diff
@@ -51,6 +52,16 @@ public AvaloniaDictionary()
         /// <inheritdoc/>
         public ICollection<TValue> Values => _inner.Values;
 
+        public bool IsFixedSize => ((IDictionary)_inner).IsFixedSize;
```
Good point, I thought that's how I'd done it tbh, but must've clicked the wrong menu item.

grokys committed:
Use discard in TryGetResource.

grokys committed:
Make `IDictionary` members explicit.

And use expression bodied members where possible.

grokys committed:
Added TryFindResource extension method.

`FindResource` can be ambiguous because it returns `AvaloniaProperty.UnsetValue` in the case of a resource not being found, which is a valid resource value. `TryFindResource` removes this ambiguity. `FindResource` has been left in for API compatibility with other frameworks.

grokys committed:
Merge branch 'master' into control-resources

grokys committed:
Prevent possible null ref warnings.

grokys committed:
Merge branch 'control-resources' of https://github.com/AvaloniaUI/Avalonia into control-resources

grokys committed:
Merge remote-tracking branch 'origin/master' into control-resources

danwalmsley reviewed (approved):
Looks good, tested with AvalonStudio, everything continued to work, and was able to migrate some of the theme to dynamic resource and change them at runtime.

grokys committed:
Merge branch 'master' into control-resources

aelij reviewed (approved):
Looks great!
